### DISCUSSION TITLE
Consider if code actions should be available in the context menu

### DISCUSSION THREAD

**Discussion by matklad** [joaotavora/eglot#1338] [Ideas]
I've long bemoaned the lack of traditional at this point :bulb: UI for executing LSP code actions:

https://rust-analyzer.github.io/blog/2020/09/28/how-to-make-a-light-bulb.html

As I've learned today, Emacs actually has a very similar UX, context menu `S-f10`! I think it would be a great re-use of existing functionality to show code actions in the context menu. Eg, here:

![image](https://github.com/joaotavora/eglot/assets/1711539/e1706ef9-ff16-4e3f-8cd8-11ab2d577ba5)

I'd love to see "extract variable"

**Comment by matklad** [comment]
Another related thing. Here's how typos look for me, using jinx:

<img width="612" height="169" alt="image" src="https://github.com/user-attachments/assets/6bed3ac9-4fbe-4ce1-86a0-da3bd62325a1" />


UI wise, this red squiggly is exactly the same kind of error visualization we get from LSP. And the action of fixing the typo is _exactly_ the same UX interaction as a code action. 

It really feels like Emacs is missing some core UX abstraction here, which would unify code actions and typo fixing under the same sort of the interface. There shouldn't be `eglot-code-actions`, there should be `execute-extended-command-at-point`, to which both eglot and jinx contribute. 


cc @joaotavora, I am going out on a limb with a ping here, but given my experience in the domain (JetBrains and rust-analyzer), this does feel like something the wider Emacs community should think about.

**Reply by joaotavora** [reply]
> It really feels like Emacs is missing some core UX abstraction here, which would unify code actions and typo fixing under the same sort of the interface. There shouldn't be eglot-code-actions, there should be execute-extended-command-at-point, to which both eglot and jinx contribute.

> cc @joaotavora, I am going out on a limb with a ping here, but given my experience in the domain (JetBrains and rust-analyzer), this does feel like something the wider Emacs community should think about.

Yes pretty firm limb and I 100% agree.  Are you sure you didn't read my rant about this yesterday?  https://lists.gnu.org/archive/html/emacs-devel/2026-08/msg00632.html

Here, to spare you the cringe, here's the paragraph I wrote about this

> Once upon a time there was a reasonably serious effort to develop a proper refactor.el here that was partially inspired by LSP and would provide a way to do much of what its code actions do. The plan was obviously that Eglot use it to link up LSP info and also allow a uniform refactor interface for other languages, including native Elisp. The entry points would be similar to Eglot's: some commands to run one or all actions, and some way to report or suggest potential actions at point (including, but obviously not limited to, diagnostic quick fixes). That effort struck the typical analysis-paralysis iceberg here in this list and I have little motivation to dig it up from the depths of the mud it has sunk into. As far as Eglot/LSP is concerned, refactorings are great and well integrated into Emacs already, even when multiple different servers are multiplexed together.

My main point is that the urgency for that super-integration has receded for me.  First, most good tools have an LSP counterpart these days. Secondly, I don't have the energy to argue through what I think is the best design.  Maybe you do?

Anyway, what LSP servers are you running here?  What is this "jinx", does it have an LSP server?   If so, merge it together with rust-analyzer or whatever Lsp server you are using.  Since you seem to like screenshots here's me running `ty`, `ruff` and `codebook-lsp serve` together:

<img width="1292" height="659" alt="image" src="https://github.com/user-attachments/assets/6de060b7-b8d0-40c3-b85a-6dbe0c857617" />

This is using [Rassumfrassum](https://github.com/joaotavora/rassumfrassum) to multiplex those three servers.  The rest of the eye candy is just the built-in Flymake's new inline diagnostics and the built-in `fido-vertical-mode`.

**Reply by matklad** [reply]
Ha, no, haven't read that, but, yeah, something like that is needed! Jinx is https://github.com/minad/jinx, the spellchecker, `M-$` thing. 

>Secondly, I don't have the energy to argue through what I think is the best design. Maybe you do?

I might have some energy, but I have absolutely no knowledge of how relevant Emacs internals work, getting back into Emacs after a decade-long detour via IntelliJ and VS Code (eglot being built in  is the reason I am back, thanks for that!)

**Reply by matklad** [reply]
Ah, codebook looks interesting! I'll probably stick with jinx for now though, as that integrates with Apple systems' spellchecker, which is a thing I think I want

**Reply by joaotavora** [reply]
I'm on holiday rn, but if you're in Lisbon in September hit me up in a DM maybe we can hack something together (and then you can try to sell it :-D )
